Perf capture minor changes for testing

kbrock opened the pull request:
pulled out of https://github.com/ManageIQ/manageiq/pull/22221

~~Mostly~~ test changes to allow the actual messages to not pay so much attention to the packaging of the collection and more attention to what is collected and the time spans.

Changes:

- ~~remove zone lookup for ems in rollup~~
- ~~remove zone lookup for gap detection alerting~~
- ~~send full 250 records to ems (was limiting vmware to 20)~~
- ~~sort batches so ids are sent more consistently~~

Fryguy commented:
Can you update the PR title to be more descriptive?  This makes it seem like it implements the full batches, but it also does other stuff including refactoring?  Hard to tell exactly what is refactor vs feature vs bug.

Fryguy commented on config/settings.yml:
```diff
@@ -883,9 +883,9 @@
-    :historical: 1
-    :hourly: 1
-    :realtime: 20
+    :historical: 250
+    :hourly: 250
+    :realtime: 250
```
This is a VMWare specific number, but it's stored here globally.  Is it possible to store it in VMware specific settings?  I wonder if we should move this section to the provider settings entirely.  @agrare Thoughts?

kbrock replied:
So, this is defined in [vmware settings](https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/config/settings.yml#L11) for:

```
:ems:
  :ems_vmware:
    :capture_batch_size: 250
```

Can we just delete these?

Up until now, we have never sent more than 1 record into the queue, so this limiter has not been used.

Now we limit the number of records going into the queue for this limit.

There is some merit to using different limits for the 3 types of collection, but it seemed like a feature that was introduced but never used.

Fryguy commented on spec/support/metric_helper.rb:
```diff
@@ -1,46 +1,54 @@
         items.each do |q|
-          obj = q.instance_id ? Object.const_get(q.class_name).find(q.instance_id) : q.class_name.constantize
+          klass = q.class_name.constantize.base_class.name
+          # exploding looks into the third argument which contains the list of object ids
+          ids = q.args[2]
```
We don't need a comment if we document the params directly in code.  I don't know what the actual params are, but something like:

```suggestion
          _param1, _param2, ids, *_ = q.args[2]
```

Fryguy replied:
(This is optional - I'm fine with `q.args[2]`...I just don't think the comment is necessary in either case.

kbrock replied:
I think naming the arguments does a great deal for documentation.

But I still find it confusing that we were using and sometimes still use `instance_id` for the target and in other instances use the third argument. So I kept the comment in there but let me know if you want me to remove it.

Fryguy commented on spec/support/metric_helper.rb:
```diff
@@ -1,46 +1,54 @@
-          messages[interval_name] ||= {}
-          (messages[interval_name][obj] ||= []) << q.args
+            messages[interval_name] ||= {}
+            (messages[interval_name][obj] ||= []) << q.args[0..1]
```
Actually, now I see you are using the first two paras here, so might be worth exploding into named params. above, then using them here.

Fryguy commented on spec/support/metric_helper.rb:
```diff
@@ -1,46 +1,54 @@
+      end
+
+      def combine_consecutive(array)
+        x = array.sort!.shift
+        array.each_with_object([]) do |i, ac|
+          # we have 2 options:
+          #   x=1-1-2000T0..23:59, i=1-2-2000T0..23:59
+          #   x=1-1-2000T0..24:00, i=1-2-2000T0..24:00
+          # so they will be within a second of eachother
+          if i.first <= (x.second + 1.second) # they overlap
+            x[1] = i.second
+          else
+            ac << x
+            x = i
+          end
+        end << x
```
Using `x.second` (i.e. the second parameter) and `1.second` (i.e. 1 time unit second) in the same line is SUPER confusing.  This might be a lot more readable with named variables...something like the below. Probably helpful to name things instead of just `x` also.

Thinking something like:

```suggestion
        prev_range_start, prev_range_end = array.sort!.shift
        array.each_with_object([]) do |(range_start, range_end), ac|
          # we have 2 options:
          #   x=1-1-2000T0..23:59, i=1-2-2000T0..23:59
          #   x=1-1-2000T0..24:00, i=1-2-2000T0..24:00
          # so they will be within a second of each other
          if range_start <= (prev_range_end + 1.second) # they overlap
            prev_range_end = i.second
          else
            ac << [prev_range_start, prev_range_end]
            range_start, range_end = prev_range_start, prev_range_end
          end
        end << [prev_range_start, prev_range_end]
```

Fryguy replied:
Now that I wrote that I think there are bugs in it?  Can you write some specs demonstrating various cases?

I'm also concerned that it modifies the incoming array with `.sort!.shift`, but that is being used in a `transform_values!` call, so I'm not sure if there are any bad side effects to that.

Fryguy replied:
Separately, asking the larger question, how do we even get times that are within a second of each other.  I thought it was built by stepping day by day and the initial day is at beginning_of_day?

kbrock replied:
sometimes we do beginning of day end of day, which is 1 second off

kbrock replied:
ASIDE: I had wanted to use a range instead of a start/end pair, but surprisingly the code was more confusing.

Fryguy commented on spec/support/metric_helper.rb:
```diff
@@ -1,46 +1,54 @@
+          # exploding looks into the third argument which contains the list of object ids
+          ids = q.args[2]
+          ids = [q.instance_id] if ids.blank?
+          objs = ids.map { |id| "#{klass}:#{id}" }.sort
```
This won't do a numerical sort if that's what you're going for?  Perhaps:

```suggestion
          objs = ids.sort.map { |id| "#{klass}:#{id}" }
```

kbrock replied:
not sure which one we want for tests. But that change looks as good as any

Fryguy commented on spec/support/metric_helper.rb:
```diff
@@ -1,46 +1,54 @@
+        "#{object.class.base_class.name}:#{object.id}"
+      end
+
+      def combine_consecutive(array)
```
Can you make this one and the one above a private method?

Fryguy commented on app/models/metric/ci_mixin/processing.rb:
```diff
@@ -14,70 +14,71 @@ def perf_process(interval_name, start_time, end_time, counters_data)
 
-  def process_counter_values(resource, data, interval_name, rt_rows)
-    log_header = "[#{interval_name}]"
+  def convert_raw_metrics_to_hashes(counters_data, interval_name)
```
If you are changing the method name, perhaps you should change the Benchmark title as well, since they were originally made to match.

kbrock replied:
I would prefer to not change this benchmark. I'll change the name back

Fryguy commented on app/models/metric/ci_mixin/processing.rb:
```diff
@@ -14,70 +14,71 @@ def perf_process(interval_name, start_time, end_time, counters_data)
-  def process_counter_values(resource, data, interval_name, rt_rows)
-    log_header = "[#{interval_name}]"
+  def convert_raw_metrics_to_hashes(counters_data, interval_name)
+    rt_rows = {}
+    counters_data.each do |resource, data|
```
Simpler?

```suggestion
    counters_data.each_with_object({}) do |(resource, data), rt_rows|
```

Fryguy commented on app/models/metric/ci_mixin/processing.rb:
```diff
@@ -95,21 +96,26 @@ def write_metrics(parameters)
   end
 
-  def update_and_publish_metrics(interval_orig, start_time, end_time, resources, rt_rows)
+  def update_last_perf_capture_on(resources, end_time)
```
These extracted methods should probably be made private.

kbrock replied:
thnx. made all of these methods private.

Fryguy commented on app/models/metric/ci_mixin/capture.rb:
```diff
@@ -90,20 +90,18 @@ def just_perf_capture(interval_name, start_time, end_time, metrics_capture)
 
     _log.info("#{log_header} Capture for #{metrics_capture.log_targets}#{log_time}...")
 
+    new_last_pref_capture_on = Time.new.utc
```
Should this be `Time.now.utc` or was this intentionally .new?

kbrock commented on app/models/metric/ci_mixin/capture.rb:
```diff
@@ -168,6 +166,10 @@ def perf_capture_callback(task_ids, _status, _message, _result)
     end
   end
 
+  def perf_capture_states(targets)
+    Benchmark.realtime_block(:capture_state) { VimPerformanceState.capture(targets) }
```
this is called from
```ruby
perf_capture_states(metrics_capture.target)
```

which is deceiving because for some metrics_capture, `target` is plural.
And tbh, going forward, more and more providers will have multiple targets (the premise of the perf_capture initiative)

kbrock commented:
updates:
- fixed Time.now
- made most of ci_mixin/processing private (all but 1 method)
- used `each_with_object`
- un-renamed `process_counter_values`
- renamed variables for test methods around historicals so it was easier to reason (i.e.: removed `i.second + 1.second`)

kbrock commented:
update:
- fixed sorting for test ids.

Fryguy commented on app/models/metric/ci_mixin/capture.rb:
```diff
@@ -90,20 +90,18 @@ def just_perf_capture(interval_name, start_time, end_time, metrics_capture)
 
     _log.info("#{log_header} Capture for #{metrics_capture.log_targets}#{log_time}...")
 
+    new_last_pref_capture_on = Time.now.utc
```
Typo?

```suggestion
    new_last_perf_capture_on = Time.now.utc
```

kbrock commented:
update:
- fixed: typo for last_pref -> last_perf

kbrock commented:
@Fryguy This was an attempt to pull out the easiest commits from #22221 - but this has stalled. This does have 9 commits, so are there any commits you want me to remove from this 
PR so we can get the easy ones merged? 

I really want to get the `config.yml` changes merged. But if taking that out of this PR means this PR gets merged soon, then just say the word.

kbrock commented:
@Fryguy @agrare Anything I can do to un-stall this?

kbrock commented:
update:
- removed non-test related issues

kicking

kbrock commented:
update:
- rebased master
- moved `combine_consecutive` to the bottom in a `private` section
